llvm/include/llvm/IR/RepoTicket.h

```diff
@@ -110,56 +115,107 @@ struct GODigestState {
   /// A reference to an object's dependencies. The GOInfo struct takes ownership
   /// of it.
   const GOVec &Dependencies;
-  /// A bool which is true if GO's hash value has been changed.
-  bool Changed;
   GODigestState() = delete;
-  GODigestState(const GOVec &Contributions, const GOVec &Dependencies,
-                bool Changed)
-      : Contributions(Contributions), Dependencies(Dependencies),
-        Changed(Changed) {}
+  GODigestState(const GOVec &Contributions, const GOVec &Dependencies)
+      : Contributions(Contributions), Dependencies(Dependencies) {}
 };
 
-/// Calculate the initial hash value, dependencies and contributions for the
-/// global object 'G'.
-/// \param G Calculated global object.
-/// \param GOI a DenseMap containing the global object information.
-///
-template <typename GlobalType>
-void calculateGOInfo(const GlobalType *G, GOInfoMap &GOI) {
-  GOInfo Result = calculateDigestAndDependenciesAndContributedToGVs(G);
-  // ContributedToGVs of an object is used to update other objects'
-  // contributions. For example, if the ContributedToGVs of function `foo`  are
-  // global variables `g` and `q`, the function `foo` is in the contributions of
-  // `g` and `q`.
-  // ContributedToGVs[`foo`] = [`g`, `q`]
-  // ====> Contributions[`g`] = [`foo`],
-  //       Contributions[`q`] = [`foo`]
-  for (auto &GO : Result.Contributions) {
-    assert(isa<GlobalVariable>(GO) &&
-           "Only global variables can have contributions!");
-    assert(isa<llvm::Function>(G) && "All contributions are functions!");
-    GOI[GO].Contributions.emplace_back(G);
+/// Helper Class for generating the hash for each global objects on the Module
+/// and it also represents how a global object (GO) is mapped into its cached
+/// hash.
+class HashCache {
+  /// A map of GO to a unique number in the dependencies graph.
+  GOStateMap Visited;
+  /// Used to keep the GO information.
+  GOInfoMap GOIMap;
+  // Used to record memoized hashes.
+  MemoizedHashes GOHashCache;
+  /// True if the module M is changed.
+  bool Changed;
+
+public:
+  HashCache() : Changed(false) {}
+  /// calculate the GO's hash and construct a memoized hashes for module.
+  HashCache(Module &M);
+
+  /// Compute the hash value for the given global object GO.
+  /// \param GO The global object.
+  /// \return The global object's digest value.
+  DigestType calculateDigest(const GlobalObject *GO);
+
+  bool isChanged() const { return Changed; }
+
+  const GOInfoMap &getGOInfoMap() const { return GOIMap; }
+
+private:
+  /// Get global object hash value from memoized hashes.
+  /// \param GO The global object.
+  /// \return A tuple of global object's hash value and  a bool which is true if
+  /// GO's hash is cached.
+  std::tuple<DigestType, bool> get(const GlobalObject *GO) const {
```

Comment:
This function name could get confused with std::get. Maybe getGOHash?

Reply:
Fixed, thanks.